File: app/src/main/java/com/example/basiccalculator/MainActivity.java
package com.example.basiccalculator;

import android.os.Bundle;
import android.text.SpannableStringBuilder;
import android.view.View;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;

import org.mariuszgromada.math.mxparser.*;

public class MainActivity extends AppCompatActivity {

    private EditText display;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        display = findViewById(R.id.input);

        display.setShowSoftInputOnFocus(false);

        display.setOnClickListener(view -> {
            if (getString(R.string.display).equals(display.getText().toString())) {
                display.setText("");
            }
        });
    }

    private void updateText(String strToAdd) {
        String oldStr = display.getText().toString();
        int cursorPos = display.getSelectionStart();
        String leftStr = oldStr.substring(0, cursorPos);
        String rightStr = oldStr.substring(cursorPos);
        if (getString(R.string.display).equals(display.getText().toString())) {
            display.setText(strToAdd);
        } else {
            display.setText(String.format("%s%s%s", leftStr, strToAdd, rightStr));
        }
        display.setSelection(cursorPos + 1);

    }

    public void zeroBTN(View view) {
        updateText("0");
    }

    public void oneBTN(View view) {
        updateText("1");
    }

    public void twoBTN(View view) {
        updateText("2");

    }

    public void threeBTN(View view) {
        updateText("3");

    }

    public void fourBTN(View view) {
        updateText("4");

    }

    public void fiveBTN(View view) {
        updateText("5");

    }

    public void sixBTN(View view) {
        updateText("6");

    }

    public void sevenBTN(View view) {
        updateText("7");

    }

    public void eightBTN(View view) {
        updateText("8");

    }

    public void nineBTN(View view) {
        updateText("9");

    }

    public void addBTN(View view) {
        updateText("+");

    }

    public void subtractBTN(View view) {
        updateText("-");

    }

    public void multiplyBTN(View view) {
        updateText("×");

    }

    public void clearBTN(View view) {
        display.setText("");

    }

    public void parenthesisBTN(View view) {
        int cursorPos = display.getSelectionStart();
        int openPar = 0;
        int closedPar = 0;
        int textLen = display.getText().length();
        for (int i = 0; i < cursorPos; i++) {
            if (display.getText().toString().charAt(i) == '(') {
                openPar += 1;
            }
            if (display.getText().toString().charAt(i) == ')') {
                closedPar += 1;
            }
        }
        if (openPar == closedPar || "+-(÷×^".indexOf(display.getText().toString().charAt(textLen - 1)) != -1) {
            updateText("(");
        } else if (closedPar < openPar && !(display.getText().toString().charAt(textLen - 1) == '(')) {
            updateText(")");

        }
        display.setSelection(cursorPos + 1);

    }

    public void exponentBTN(View view) {
        updateText("^");

    }

    public void divideBTN(View view) {
        updateText("÷");

    }

    public void pointBTN(View view) {
        updateText(".");

    }

    public void backspaceBTN(View view) {
        int cursorPos = display.getSelectionStart();
        int textLength = display.getText().length();

        if (cursorPos != 0 && textLength != 0) {
            SpannableStringBuilder selection = (SpannableStringBuilder) display.getText();
            selection.delete(cursorPos - 1, cursorPos);
        }
    }

    public void equalsBTN(View view) {
        String userExp = display.getText().toString();
        userExp = userExp.replaceAll("÷", "/");
        userExp = userExp.replaceAll("×", "*");
        Expression exp = new Expression(userExp);
        String result = String.valueOf(exp.calculate());
        display.setText(result);
        display.setSelection(result.length());
    }

    public void plusMinusBTN(View view) {
        updateText("-");

    }
}

File: app/src/main/java/com/example/basiccalculator/MainActivity.kt
package com.example.basiccalculator

import androidx.appcompat.app.AppCompatActivity
import android.widget.EditText
import android.os.Bundle
import android.text.SpannableStringBuilder
import android.view.View
import org.mariuszgromada.math.mxparser.Expression

class MainActivity : AppCompatActivity() {
    private lateinit var display: EditText
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        display = findViewById(R.id.value)
        display.showSoftInputOnFocus = false
    }

    private fun updateText(strToAdd: String) {
        val oldStr = display.text.toString()
        val cursorPos = display.selectionStart
        val leftStr = oldStr.substring(0, cursorPos)
        val rightStr = oldStr.substring(cursorPos)
        display.setText(String.format("%s%s%s", leftStr, strToAdd, rightStr))
        display.setSelection(cursorPos + 1)
    }

    fun zeroBTN(view: View?) {
        updateText("0")
    }

    fun oneBTN(view: View?) {
        updateText("1")
    }

    fun twoBTN(view: View?) {
        updateText("2")
    }

    fun threeBTN(view: View?) {
        updateText("3")
    }

    fun fourBTN(view: View?) {
        updateText("4")
    }

    fun fiveBTN(view: View?) {
        updateText("5")
    }

    fun sixBTN(view: View?) {
        updateText("6")
    }

    fun sevenBTN(view: View?) {
        updateText("7")
    }

    fun eightBTN(view: View?) {
        updateText("8")
    }

    fun nineBTN(view: View?) {
        updateText("9")
    }

    fun addBTN(view: View?) {
        updateText("+")
    }

    fun subtractBTN(view: View?) {
        updateText("-")
    }

    fun multiplyBTN(view: View?) {
        updateText("×")
    }

    fun clearBTN(view: View?) {
        display.setText("")
    }

    fun parenthesisBTN(view: View?) {
        val cursorPos = display.selectionStart
        var openPar = 0
        var closedPar = 0
        val textLen = display.text.length
        for (i in 0 until cursorPos) {
            if (display.text.toString()[i] == '(') {
                openPar += 1
            }
            if (display.text.toString()[i] == ')') {
                closedPar += 1
            }
        }
        if (openPar == closedPar || "+-(÷×^".indexOf(display.text.toString()[textLen - 1]) != -1) {
            updateText("(")
        } else if (closedPar < openPar && display.text.toString()[textLen - 1] != '(') {
            updateText(")")
        }
        display.setSelection(cursorPos + 1)
    }

    fun exponentBTN(view: View?) {
        updateText("^")
    }

    fun divideBTN(view: View?) {
        updateText("÷")
    }

    fun pointBTN(view: View?) {
        updateText(".")
    }

    fun backspaceBTN(view: View?) {
        val cursorPos = display.selectionStart
        val textLength = display.text.length
        if (cursorPos != 0 && textLength != 0) {
            val selection = display.text as SpannableStringBuilder
            selection.delete(cursorPos - 1, cursorPos)
        }
    }

    fun equalsBTN(view: View?) {
        var userExp = display.text.toString()
        userExp = userExp.replace("÷".toRegex(), "/")
        userExp = userExp.replace("×".toRegex(), "*")
        val exp = Expression(userExp)
        val result = exp.calculate().toString()
        display.setText(result)
        display.setSelection(result.length)
    }

    fun plusMinusBTN(view: View?) {
        updateText("-")
    }
}